[For Review] Menu/Popover/IconMenu/Dialog Portal updates

chrismcv opened the pull request:
This PR 
a) currently breaks dialogs
b) introduces a component `Popover` which is an arbitrary control that shows a popup menu, either `onClick` or on `props.open = true`.
c) Popover component wraps menu in IconMenu - nearly transparently.
d) Popover renders to a new DOM Tree, and in theory context should transfer for React@0.13
e) Dialog renders into new tree, but has issues that I'll sort shortly...

patrykkopycinski commented on docs/src/app/app-routes.jsx:
```diff
@@ -67,7 +68,7 @@ let AppRoutes = (
     <Route name="customization" handler={Customization}>
       <Route name="colors" handler={Colors} />
-      <Route name="themes" handler={Themes} />
+      <Route name="themes" handler={Themes} />let
```
is it necessary?

chrismcv commented:
Updated PR:
 - Dialog is fixed via a wrapper
 - I'm having a problem with the DatePickerDialog/TimePickerDialog and contexts - any suggestions welcome @hai-cea @shaurya947 
 - I think there is some scope to simplify menu's now that animation/positioning is delegated to the popover component

chrismcv commented:
I've updated this PR - lots of changes....

- New component "Popover"
- New mixin "Render to layer"
- React 14 support 

Updated the following components to use it
- Dropdown menu
- Icon Menu
- Dialog
- SelectField
- Menu (menus/menu-item.jsx) when it has nested items

It'd be great if review + merge process could start?

oliviertassinari commented on src/text-field.jsx:
```diff
@@ -331,7 +331,8 @@ const TextField = React.createClass({
     }
     if (this.props.children) {
-      inputElement = React.cloneElement(this.props.children, {...inputProps, ...this.props.children.props});
+      let childInputStyle = this.mergeStyles(inputStyle, this.props.children.style);
```
Could you remove the extra commit from https://github.com/callemall/material-ui/pull/1896?

chrismcv replied:
#1896 is that single commit... do you want it removed from #1845 (the big one?)

oliviertassinari replied:
Well, I think that #1896 will be reviewed and merged before this PR. But as you want

shaurya947 commented:
@chrismcv are you still having the warnings regarding `context`? I was hoping they would go away with React v14 as `context` moves from being owner-based to parent-based.

@oliviertassinari seems like you have reviewed this a little. Is it merge-ready?

chrismcv commented:
v14 has sorted context issues happily.  I removed the initial context hacks that I had in for v13.

I think the changes here are significant, and technically, they are just about there now.

There are a few more docs changes I think we should put in. (And at least one fix to docs site "customisation section").

I've made dialog have an `open` property which controls visibility and depreciates `show` and `hide`.  There is a manual override that allows `show` and `hide` but this is defaulted to `false` which means a breaking change.

The Dropdown/Icon Menu transitions are subtly different from previous.

The bit I'd like you to look at is the interface for position the popover.  I've introduced `anchorEl` and `targetEl` positions.  So you can say that the top left of the popover should show originating at the bottom right of the target element.  (see icon menu docs for examples).

My nested menu support is a little rudimentary, its currently missing some keyboard support - but this seems partly true of the new menus stuff in general.

shaurya947 commented:
@chrismcv look at #1996 regarding props for `Dialog`.

oliviertassinari commented:
@shaurya947 not yet

chrismcv commented:
@shaurya947 I've landed on something very similar to #1996, bar my property is called `open` rather than `isOpen` and I haven't removed the imperative API, though am happy to do that?  Was that the final agreement for #1996?

shaurya947 commented:
@chrismcv we're slightly short of a *final agreement* on that. Should reach a conclusion soon hopefully! :-)

oliviertassinari commented on docs/src/app/components/pages/components/popover.jsx:
```diff
@@ -0,0 +1,157 @@
+        ],
+      },
+      {
+        name: 'Methods',
```
Do we need those imperative methods? I would rather avoid them.

oliviertassinari commented on src/mixins/render-to-layer.js:
```diff
@@ -0,0 +1,119 @@
+import debounce from 'lodash.debounce';
+
+// heavily inspired by https://github.com/Khan/react-components/blob/master/js/layered-component-mixin.jsx
+const LayerMixin = {
```
Can't this be a component instead?

oliviertassinari commented on src/time-picker/clock-hours.jsx:
```diff
@@ -32,6 +32,9 @@ const ClockHours = React.createClass({
     muiTheme: React.PropTypes.object,
   },
 
+    format: React.PropTypes.oneOf(['ampm', '24hr']),
```
That's weird

chrismcv replied:
Agreed - seems to be missing `propTypes` from a rebase...

oliviertassinari commented:
@chrismcv There is a lot in this PR. Can't we split this with independent bit?

chrismcv commented:
@oliviertassinari - yes I agree.  
Happy to split - and can do so this evening I think.

Does the following make sense (1+2 being dependencies for 3+4)

1. Render to Layer component
2. Popover component
3. Dialog to use render to layer (and fix other dialogs + docs)
4. IconMenu/NestedMenus

shaurya947 commented:
@chrismcv is this PR stale then? Are you going to close it?

chrismcv commented:
closing for now - more to follow
